### ISSUE TITLE
"Maximum call stack size exceeded" on build

### ISSUE THREAD

**Issue by ismyrnow** [stefanpenner/ember-fetch#7]
My ember-cli builds are failing since introducing ember-fetch as a dependency.

This may be an issue with broccoli-stew, but hopefully the stack trace below will help pinpoint the exact issue.

I'm on Windows 8.1 using ember-cli v0.2.3.

I know this version of ember-cli is out of date. I can update my global cli version, but updating the app isn't an option.

```
Build failed.
Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at new Error (native)
    at Error (native)
    at Object.fs.mkdirSync (fs.js:799:18)
    at sync (C:\Users\ismyrnow\Git\myapp\dashboard\node_modules\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:71:13)
    at sync (C:\Users\ismyrnow\Git\myapp\dashboard\node_modules\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:77:24)
    at sync (C:\Users\ismyrnow\Git\myapp\dashboard\node_modules\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:78:17)
    at sync (C:\Users\ismyrnow\Git\myapp\dashboard\node_modules\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:78:17)
    at sync (C:\Users\ismyrnow\Git\myapp\dashboard\node_modules\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:78:17)
    at sync (C:\Users\ismyrnow\Git\myapp\dashboard\node_modules\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:78:17)
    at sync (C:\Users\ismyrnow\Git\myapp\dashboard\node_modules\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:78:17)
```

**Comment by knownasilya** [comment]
On OSX I don't see this error, but I do see triples of every build.

**Comment by stefanpenner** [comment]
> I'm on Windows 8.1 using ember-cli v0.2.3.

this version of ember-cli is quite old, and not supported. I would recommend upgrading.

> but I do see triples of every build.

What does this mean? Maybe this deserves its own issue

**Comment by knownasilya** [comment]
Ok, seems like it was solved some how (restart of laptop), maybe it was a watchman issue. Ignore my comment.

**Comment by stefanpenner** [comment]
> Ok, seems like it was solved some how (restart of laptop), maybe it was a watchman issue. Ignore my comment.

:+1: I suspect this is problem you mention will go away once i land the next batch of broccoli perf work. Sorry for any interim pains.

**Comment by ismyrnow** [comment]
We upgraded our app to the latest ember-cli, and I am still unable to build the project. This time, with a different error message:

```
> ember build
version: 1.13.8
Build failed.
EEXIST: file already exists, mkdir 'C:\Users\ismyrnow\Git\myapp\dashboard\tmp\funnel-dest_8buALG.tmp'
```

I have tried the following:

- Deleting my tmp and dist folders before building.
- Reinstalling all dependencies (including cleaning my npm and bower cache, as described in the ember-cli update docs).
- Reinstalling my global ember-cli module.
- Rebooting my machine, and attempting to build immediately after startup.

When I remove the ember-fetch dependency, and comment out the dependent code, the build completes successfully.

**Comment by calvinmetcalf** [comment]
if I put a console.log [here](https://github.com/stefanpenner/ember-fetch/blob/15fca46eb7c55458063d06ddc7018d9e69c01381/index.js#L28) for 'fetch' I also see that error, but on a mac it apparently ignore it?

**Comment by ismyrnow** [comment]
I cloned this repo and tried to run the tests locally, and I received the same error message as before.

Note that this project depends on ember-cli@0.2.3 for tests, which is the same version we were using previously, and the error message is the same as I encountered previously.

```
> npm test

> ember-fetch@1.2.2 test C:\ember-fetch
> ember test

Future versions of Ember CLI will not support v3.3.1. Please update to Node 0.12 or io.js.
version: 0.2.3
1.13.8

Build failed.
Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at new Error (native)
    at Error (native)
    at Object.fs.mkdirSync (fs.js:799:18)
    at sync (C:\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:71:13)
    at sync (C:\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:77:24)
    at sync (C:\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:78:17)
    at sync (C:\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:78:17)
    at sync (C:\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:78:17)
    at sync (C:\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:78:17)
    at sync (C:\ember-fetch\node_modules\broccoli-stew\node_modules\broccoli-funnel\node_modules\mkdirp\index.js:78:17)
```

**Comment by ismyrnow** [comment]
Some poking around shows that the issue is when calling `find(expandedFetchPath)` [here](https://github.com/stefanpenner/ember-fetch/blob/master/index.js#L25).

I suspect that some of the tools in `broccoli-stew` aren't windows compatible due to using `/` for path separators. When I use `path.sep` in the `index.js` file of this project, the underlying modules still seem to throw errors.

For example, when calling `find('C:\ember-fetch\node_modules\whatwg-fetch\{fetch.js}')`, I get the following:

```
top level glob or expansion not currently supported: `C:\ember-fetch\node_modules\whatwg-fetch\{fetch.js}`
Error: top level glob or expansion not currently supported: `C:\ember-fetch\node_modules\whatwg-fetch\{fetch.js}`
    at Error (native)
    at C:\ember-fetch\node_modules\broccoli-stew\lib\utils\decompose.js:34:83
    at Array.forEach (native)
    at module.exports (C:\ember-fetch\node_modules\broccoli-stew\lib\utils\decompose.js:14:19)
    at _find (C:\ember-fetch\node_modules\broccoli-stew\lib\find.js:106:25)
    at C:\ember-fetch\node_modules\broccoli-stew\lib\find.js:97:47
    at find (C:\ember-fetch\node_modules\broccoli-stew\lib\find.js:91:54)
    at Class.module.exports.treeForVendor (C:\ember-fetch\index.js:27:46)
    at Class._treeFor (C:\ember-fetch\node_modules\ember-cli\lib\models\addon.js:315:31)
    at Class.treeFor (C:\ember-fetch\node_modules\ember-cli\lib\models\addon.js:285:19)
```

**Comment by stefanpenner** [comment]
@ismyrnow that looks like a broccoli-stew bug, it shouldn't be passing through the `{`. Maybe its something funky re: windows paths `\`

**Comment by ismyrnow** [comment]
I tried updating the path building code to use system path separators (`path.sep`), but it didn't make a difference. I got a different error, but as I recall, similar stack traces.

**Comment by stefanpenner** [comment]
many related bugs have been fixed on broccoli-stew, if the related issue is not fixed on the latest version, will gladly reopen.

**Comment by ismyrnow** [comment]
Did fresh install of all dependencies (including cache cleaning) and I'm still getting similar errors.

Without any changes, I get a `EEXIST: file already exists` error during ember build.

My attempts at tracking this down have been unfruitful. I agree that it may be related to path separators on Windows. Digging into various dependencies (`broccoli-stew`, `broccoli-templater`,and `broccoli-funnel`) I see lots of hard-coded path separators.

That said, it may be easier to track this down by testing those modules mentioned above on Windows. I don't have the time to do that at the moment, nor do I know if tests in those modules pass on Windows.

**Comment by DLiblik** [comment]
This comment is a NOOP, but just for further affirmation, as a safety rule we regex \ to / as an input guard on any Windows-hosted broccolis we run as the good old Windows backslash seems to cause rabbit holes that were mysteriously claiming devs on our team. Given broccoli's focus on paths, I think this needs to be solved in the bedrock.

**Comment by stefanpenner** [comment]
> Digging into various dependencies (broccoli-stew, broccoli-templater,and broccoli-funnel) I see lots of hard-coded path separators.

there is nothing wrong with hard-coded path separators, all fs operations normalize before doing actual work. Like we are doing some comparison somewhere, causing grief.

**Comment by DLiblik** [comment]
If it helps, we did it to ourselves most often on some of our own custom nodes when we used .split('/') to segment a path when processing custom steps during ember-cli builds - that was our most common usage of explicit paths (versus fs doing the normalization in other cases as @stefanpenner points out).  For anyone desperate for a workaround, explicitly normalizing input path variables with .replace(/\\/g,'/') might get you there.

**Comment by stefanpenner** [comment]
For reference, I have been going around doing: https://github.com/joliss/node-walk-sync/blob/master/index.js#L26-L32

**Comment by DLiblik** [comment]
That's a better approach - deleting my prior comment.

**Comment by stefanpenner** [comment]
making it a node_module \w tests now

**Comment by stefanpenner** [comment]
https://github.com/stefanpenner/ensure-posix-path

**Comment by stefanpenner** [comment]
looking into this one now, have a failing test on broccoli-stew. Going to add some more related win32 tests.

Going forward, it would be quite helpful if failing test cases be provided.

**Comment by stefanpenner** [comment]
I have fixed all issues I can find in broccoli-stew (related to this issue), there appears to be a broccoli-funnel issue. I hope to dig into that one next.

Stay tunned.

**Comment by stefanpenner** [comment]
I believe I have addressed the issue [here](https://github.com/stefanpenner/broccoli-stew/pull/96) stew now includes this use-case explicitly in its test suite https://github.com/stefanpenner/broccoli-stew/pull/96/files#diff-bb2e0e63dee5aa59701fe153f0a81d0eR297

I'm going to add windows CI to this project before closing the issue

**Comment by ismyrnow** [comment]
If I update the "broccoli-stew" dependency to "^1.2.0", I am able to successfully build on my Windows machine. Do you suspect any breaking changes to this project for switching from 0.x to 1.x there?

**Comment by stefanpenner** [comment]
1.3.0 should fix this (just tested in windows, seems to do the right thing)

**Comment by ismyrnow** [comment]
Just confirmed that I can successfully run `npm test` for this project. The output is bower install stuff, ember build, then tests run with phantomjs.

On a side note, the tests seem to run indefinitely. It looped through at least 3 install/build/test iterations before I shut it down. Seems more like an ember-cli issue, though.

**Comment by knownasilya** [comment]
Tests run thorough ember-try (https://github.com/stefanpenner/ember-fetch/blob/master/config/ember-try.js) which tests a matrix of ember versions.

**Comment by stefanpenner** [comment]
> On a side note, the tests seem to run indefinitely. It looped through at least 3 install/build/test iterations before I shut it down. Seems more like an ember-cli issue, though.

sounds like you need to be more patient, the tests run many permutations.

**Comment by ismyrnow** [comment]
Didn't know what ember-try was doing. Thanks for the info.
